src/composite_systems.jl

```diff
@@ -0,0 +1,157 @@
+using LinearAlgebra: I
+import TensorOperations
+
+export CompositeQSystem, add_hamiltonian!, add_lindblad!
+
+# tensor products of quantum systems
+mutable struct CompositeQSystem
+    subsystems::Vector
+    fixed_Hs::Vector{Tuple} # tuple of Matrix and expansion indices
+    parametric_Hs::Vector{Tuple} # tuple of Functions and expansion indices
+    fixed_Ls::Vector{Tuple} # tuple of Matrix and expansion indices for collapse operators
+    parametric_Ls::Vector{Tuple} # tuple of Matrix, expansion indices, and time-dependent function for collapse operators
+    dimension::Int
+end
+
+CompositeQSystem(qs) = CompositeQSystem(qs, [], [], [], [], prod(dimension(q) for q in qs))
+
+# helper functions for CompositeQSystems
+dimension(cqs::CompositeQSystem) = cqs.dimension
+dim(cqs::CompositeQSystem) = dimension(cqs)
+
+find_indices(cqs::CompositeQSystem, s::QSystem) = findall([sub == s for sub in cqs.subsystems])
+find_indices(cqs::CompositeQSystem, s_label::AbstractString) = findall([label(sub) == s_label for sub in cqs.subsystems])
+function find_indices(cqs::CompositeQSystem, s::Union{Vector{<:QSystem}, Vector{<:AbstractString}})
+    return vcat([find_indices(cqs, _s) for _s in s]...)
+end
+
+""" Add a subsystem static Hamiltonian matrix to a CompositeQSystem """
+function add_hamiltonian!(cqs::CompositeQSystem, ham::AbstractMatrix{T}, acting_on::Union{Q, Array{Q}}) where {T<:Number, Q<:QSystem}
```

Comment:
Not that it matters, but since `T` isn't referred to in the method, couldn't you write `ham::AbstractMatrix{<:Number}`?